### PR TITLE
BUG: group exprdata by the expressions, not identity

### PR DESCRIPTION
**Pull request by llllllllll**
When comparing exprdata, check `isidentical` on the constituent expressions instead of identity on the exprdata tuple itself. There are cases where we don't register the whole underlying table in a single call to `register_dataset` because it is difficult to group the columns this way. The equality of the exprdata object determines how we batch queries to the underlying, so we want to group them up as much as possible or we will be querying for the same `sid`, `asof_date`, and `timestamp` columns and re-aligning them for multiple columns.

### CODE REVIEW

File: zipline/pipeline/loaders/blaze/core.py [hunk 1]
```diff
@@ -729,25 +729,36 @@ def __new__(cls,
...
     def __eq__(self, other):
-        return self is other
+        if not isinstance(other, ExprData):
+            return NotImplemented
+
+        return (
+            self.expr.isidentical(other.expr) and
+            self._expr_eq(self.deltas, other.deltas) and
+            self._expr_eq(self.checkpoints, other.checkpoints) and
+            self.odo_kwargs == other.odo_kwargs
```
Comment (ssanderson) [inline zipline/pipeline/loaders/blaze/core.py]: Any worry that `odo_kwargs` might contain object that aren't comparable and/or objects that don't have the equality semantics we want?

File: zipline/pipeline/loaders/blaze/core.py [hunk 2]
```diff
@@ -729,25 +729,36 @@ def __new__(cls,
...
+        )
+
+    def __ne__(self, other):
```
Comment (ssanderson) [inline zipline/pipeline/loaders/blaze/core.py]: I think this is the default behavior. Is there a reason to implement this explicitly?

```
In [16]: class Foo(object):
    ...:
    ...:     def __eq__(self, other):
    ...:         print("calling eq")
    ...:         return True
    ...:
    ...:

In [17]: Foo() == Foo()
calling eq
Out[17]: True

In [18]: Foo() != Foo()
calling eq
Out[18]: False
```

File: zipline/pipeline/loaders/blaze/core.py [hunk 3]
```diff
@@ -729,25 +729,36 @@ def __new__(cls,
...
-            odo_kwargs or {},
+            frozenset((odo_kwargs or {}).items()),
         )
 
     def __repr__(self):
         # If the expressions have _resources() then the repr will
         # drive computation so we take the str here.
-        cls = type(self)
-        return super(ExprData, cls).__repr__(cls(
+        return repr(_expr_data_base(
```
Comment (ssanderson) [inline zipline/pipeline/loaders/blaze/core.py]: Why this change over the `super`-based implementation?

File: zipline/pipeline/loaders/blaze/core.py [hunk 4]
```diff
@@ -729,25 +729,36 @@ def __new__(cls,
...
+    def _expr_eq(a, b):
+        return a is b is None or a.isidentical(b)
 
     def __eq__(self, other):
-        return self is other
+        if not isinstance(other, ExprData):
+            return NotImplemented
+
+        return (
+            self.expr.isidentical(other.expr) and
```
Comment (ssanderson) [inline zipline/pipeline/loaders/blaze/core.py]: Are we not using `_expr_eq` here to avoid the extra cost of the two `is None` checks? That seems unlikely to be worth the cost over the improved clarity of making it clear that we want to do the logical comparison for all three blaze expressions. At the very least, we should comment here that these are the same logical check (assuming that's the case and I'm not missing something). If the perf does matter that much, should we be binding `_expr_eq` as a local?

File: zipline/pipeline/loaders/blaze/core.py [hunk 5]
```diff
@@ -729,25 +729,36 @@ def __new__(cls,
             expr,
             deltas,
             checkpoints,
-            odo_kwargs or {},
+            frozenset((odo_kwargs or {}).items()),
```
Comment (ssanderson) [inline zipline/pipeline/loaders/blaze/core.py]: Why the change from dict to frozenset? Is this just being paranoid about mutation? Are all the objects we pass as values to `odo_kwargs` hashable?

File: zipline/pipeline/loaders/blaze/core.py (hunk 4 above)
Comment (llllllllll) [reply zipline/pipeline/loaders/blaze/core.py]: `expr` can't be None so I didn't consider using that. I could use the method to be more clear though.

File: zipline/pipeline/loaders/blaze/core.py (hunk 5 above)
Comment (llllllllll) [reply zipline/pipeline/loaders/blaze/core.py]: They are currently hashable (bytes values, unicode keys). This is because we want the `__hash__` of `ExprData` to be by value, not identity.

File: zipline/pipeline/loaders/blaze/core.py (hunk 1 above)
Comment (llllllllll) [reply zipline/pipeline/loaders/blaze/core.py]: No, these are mostly simple objects like strings, bytes, or ints. I think the only real use site we have for this is `bytes`.

File: zipline/pipeline/loaders/blaze/core.py (hunk 2 above)
Comment (llllllllll) [reply zipline/pipeline/loaders/blaze/core.py]: I believe `tuple` changes this:
```python
In [9]: class T(tuple):
   ...:     def __eq__(self, other):
   ...:         print('eq')
   ...:         return super(T, self).__eq__(other)
   ...:     
   ...:     

In [10]: t = T((1, 2, 3))

In [11]: t == t
eq
Out[11]: True

In [12]: t != t
Out[12]: False

```

File: zipline/pipeline/loaders/blaze/core.py (hunk 3 above)
Comment (llllllllll) [reply zipline/pipeline/loaders/blaze/core.py]: I actually want to call the super class's constructor and then repr that. I don't want to do the `dict` to `frozenset` conversion. This seemed like the most straightforward way to do that. I am not that worried about some complex hierarchy below this type.

File: zipline/pipeline/loaders/blaze/core.py (hunk 2 above)
Comment (ssanderson) [reply zipline/pipeline/loaders/blaze/core.py]: Ah. Can we add a comment here noting that we need a `__ne__` to override the one we get from being a tuple subclass?
